README
MooseX::Accessors::ReadWritePrivate version 1.3.0.





INSTALLATION

To install this module, ensure Module::Build is installed, then run the
following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install



NOTE ON ENCODINGS

All files in this distribution should be considered to be in UTF-8 format.
Some standard documentation tools do not presently deal very well with this and
their output is mangled.  Give that it is the 21st century, this is considered
a bug in those tools and not in this distribution.



COPYRIGHT AND LICENCE

Copyright ©2009-2010, Elliot Shank <perl@galumph.com>. All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for MooseX::Accessors::ReadWritePrivate.

v1.3.0  2010/7/21

    Support non-public selectors.

    Deal with deprecations in Moose.


v1.2.1  2009/7/26

    Change minimum Moose version requirement to 0.84 in order to allow for
    "bare" in the tests.


v1.2.0  2009/7/20

    Handle "is => 'bare'" attributes.


v1.1.0  2009/4/20

    Treat Maybe[Bool] attributes the same way as Bool ones.


v1.0.1  2009/4/18

    Ack!  Listed dependencies were completely wrong.


v1.0.0  2009/4/18

    Initial release.


lib/MooseX/Accessors/ReadWritePrivate.pm
package MooseX::Accessors::ReadWritePrivate;

use 5.008;
use utf8;
use strict;
use warnings;

use version; our $VERSION = qv('v1.3.0');

use Moose 0.94 qw< >;
use Moose::Exporter;
use Moose::Util::MetaRole;

use MooseX::Accessors::ReadWritePrivate::Role::Attribute;


Moose::Exporter->setup_import_methods(
    class_metaroles => {
        attribute => [
            qw< MooseX::Accessors::ReadWritePrivate::Role::Attribute >
        ],
    },
);


1;

__END__

=pod

=encoding utf8

=for stopwords ro rwp

=head1 NAME

MooseX::Accessors::ReadWritePrivate - Name your accessors get_foo() and set_foo() or _set_foo().


=head1 SYNOPSIS

    package Some::Class;

    use Moose;
    use MooseX::Accessors::ReadWritePrivate;

    has foo => {
        is => 'rw',  # Will create get_foo() and set_foo().
        ...
    };
    has _foo => {
        is => 'rw',  # Will create _get_foo() and _set_foo().
        ...
    };
    has __foo => {
        is => 'rw',  # Will create __get_foo() and __set_foo().
        ...
    };
    has bar => {
        is => 'ro',  # Will create get_bar().
        ...
    };
    has bar => {
        is => 'rop',  # Will create _get_bar().
        ...
    };
    has baz => {
        is => 'rpwp', # Will create _get_baz() and _set_baz().
        ...
    };
    has _baz => {
        is => 'rpwp', # Will create _get_baz() and _set_baz().
        ...
    };
    has __baz => {
        is => 'rpwp', # Will create _get_baz() and _set_baz().
        ...
    };

    has special_reader => {
        is      => 'rw',
        reader  => 'blah_blah', # Will still create set_special_reader().
        ...
    };
    has special_writer => {
        is      => 'rw',
        writer  => 'blah_blah', # Will still create get_special_writer().
        ...
    };

    has is_hot => {
        isa => 'Bool',
        is  => 'rw',    # Will create is_hot() and set_is_hot().
        ...
    };
    has is_cold => {
        isa => 'Bool',
        is  => 'ro',    # Will create is_cold().
        ...
    };
    has is_tepid => {
        isa => 'Bool',
        is  => 'rwp',   # Will create is_tepid() and _set_is_tepid().
        ...
    };


=head1 VERSION

This document describes MooseX::Accessors::ReadWritePrivate version 1.3.0.


=head1 DESCRIPTION

This module does not provide any methods.  Simply loading it changes the
default naming policy for the loading class so that accessors are separated
into a selector and a mutator.

Assuming that there is no "p" or "wo" in the C<is> declaration, the selector
will be named the same as the attribute with "get_" prefixed, unless the
attributes is a C<Bool> or a C<Maybe[Bool]>, in which case the selector will
have the same name as the attribute.

If there is no "ro", in the C<is> declaration, a mutator will be crated.
Assuming that there is no "p" in the declaration, the mutator will be named
the same as the attribute with "set_" prefixed, modulo leading underscores.

Appending a "p" in the C<is> declaration after a given component indicates
that you want the accessor "private" (as much as anything in Perl is private),
by having a single leading underscore in the name, regardless of the number of
leading underscores in the attribute name.  For example, using "rpw" will make
the selector private and the mutator private and using "rop" will make the
selector private and not create a mutator.

If the attribute has any leading underscores, those will moved to the front of
the accessor names, unless the accessor is supposed to be private, in which
case the mutator will have a single underscore at the front.

To be absolutely plain, here's the complete list of acceptable values for C<is>:

=over

=item rw

=item ro

=item wo

=item rpw

=item rwp

=item rpwp

=item rop

=item wop

=item bare

=back


Unlike other accessor naming schemes for L<Moose>, specifying a value for the
"reader" only or the "writer" only does not disable the behavior of this
module for the other accessor.


=head1 DIAGNOSTICS

None.


=head1 CONFIGURATION AND ENVIRONMENT

None other than what you specify for your attributes.


=head1 DEPENDENCIES

perl 5.8

L<Moose> 0.56

L<Moose::Exporter>

L<Moose::Util::MetaRole>

L<version>


=head1 AUTHOR

Elliot Shank C<< <perl@galumph.com> >>


=head1 LICENSE AND COPYRIGHT

Copyright ©2009-2010, Elliot Shank C<< <perl@galumph.com> >>.

Based upon L<MooseX::FollowPBP>, copyright ©2008 Dave Rolsky.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE
SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE
STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE
SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND
PERFORMANCE OF THE SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE,
YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY
COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE
SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE LIABLE TO YOU FOR DAMAGES,
INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING
OUT OF THE USE OR INABILITY TO USE THE SOFTWARE (INCLUDING BUT NOT LIMITED TO
LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR
THIRD PARTIES OR A FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER
SOFTWARE), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGES.

=cut


# setup vim: set filetype=perl tabstop=4 softtabstop=4 expandtab :
# setup vim: set shiftwidth=4 shiftround textwidth=78 autoindent :
# setup vim: set foldmethod=indent foldlevel=0 :


lib/MooseX/Accessors/ReadWritePrivate/Role/Attribute.pm
package MooseX::Accessors::ReadWritePrivate::Role::Attribute;

use 5.008;
use utf8;

use strict;
use warnings;


use version; our $VERSION = qv('v1.3.0');


use Moose::Role;


my $TRUE  = 1;
my $FALSE = 0;


before _process_options => sub {
    my ($class, $name, $options) = @_;

    my $is = $options->{is};
    if (
            $is
        and not ( exists $options->{reader} and exists $options->{writer} )
        and $is =~ m<
            \A
            (?:
                    r (p)? w (p)?
                |   ( [rw] ) o (p)?
            )
            \z
        >xms
    ) {
        my ($read_private, $write_private, $read_or_write_only, $only_private) =
            ($1, $2, $3, $4);

        if (
            $name =~ m<
                \A
                ( _* )
                ( [^_] .* )?
                \z
            >xms
        ) {
            my ($scope, $base) = ($1, $2);

            my $generate_selector = not exists $options->{reader};
            my $generate_mutator = not exists $options->{writer};
            if ($read_or_write_only) {
                if ($read_or_write_only eq 'r') {
                    $read_private = $only_private;
                    $generate_mutator = $FALSE;
                } else {
                    $write_private = $only_private;
                    $generate_selector = $FALSE;
                } # end if
            } # end if

            if ($generate_selector) {
                my $read_scope = $read_private ? '_' : $scope;

                my $type = $options->{isa};
                my $prefix;
                if ( $type and ($type eq 'Bool' or $type eq 'Maybe[Bool]') ) {
                    $prefix = q<>;
                } else {
                    $prefix = 'get_';
                } # end if

                $options->{reader} = "$read_scope$prefix$base";
            } # end if

            if ($generate_mutator) {
                if ($write_private) {
                    $options->{writer} = "_set_$base";
                } else {
                    $options->{writer} = "${scope}set_$base";
                } # end if
            } # end if

            delete $options->{is};
        } # end if
    } # end if

    return;
}; # end before _process_options()


no Moose::Role;

1;

__END__

=pod

=encoding utf8

=for stopwords

=head1 NAME

MooseX::Accessors::ReadWritePrivate::Role::Attribute - Names (non Bool) accessors affordance style.


=head1 SYNOPSIS

None.  This is part of the implementation of L<MooseX::Accessors::ReadWritePrivate>.


=head1 VERSION

This document describes MooseX::Accessors::ReadWritePrivate::Role::Attribute
version 1.3.0.


=head1 DESCRIPTION

This role applies a method modifier to the C<_process_options()>
method, and tweaks the reader and writer parameters so that they
end up with affordance names, unless they are C<Bool>s, in which case the
names are semi-affordance.


=head1 DIAGNOSTICS

None.


=head1 CONFIGURATION AND ENVIRONMENT

None other than what you specify for your attributes.


=head1 DEPENDENCIES

perl 5.8

L<Moose::Role>

L<version>


=head1 AUTHOR

Elliot Shank C<< <perl@galumph.com> >>


=head1 LICENSE AND COPYRIGHT

Copyright ©2009-2010, Elliot Shank C<< <perl@galumph.com> >>.

Based upon L<MooseX::FollowPBP>, copyright ©2008 Dave Rolsky.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.


=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE
SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE
STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE
SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND
PERFORMANCE OF THE SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE,
YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY
COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE
SOFTWARE AS PERMITTED BY THE ABOVE LICENSE, BE LIABLE TO YOU FOR DAMAGES,
INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING
OUT OF THE USE OR INABILITY TO USE THE SOFTWARE (INCLUDING BUT NOT LIMITED TO
LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR
THIRD PARTIES OR A FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER
SOFTWARE), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGES.

=cut


# setup vim: set filetype=perl tabstop=4 softtabstop=4 expandtab :
# setup vim: set shiftwidth=4 shiftround textwidth=78 autoindent :
# setup vim: set foldmethod=indent foldlevel=0 :


Build.PL
use strict;
use warnings;

use English qw< -no_match_vars >;

use Module::Build;


my $class = Module::Build->subclass( code => <<'END_SUBCLASS' );

    sub ACTION_authortest {
        my ($self) = @_;

        $self->depends_on('build');
        $self->depends_on('manifest');
        $self->depends_on('distmeta');

        $self->test_files( qw< t xt/author > );
        $self->recursive_test_files(1);

        $self->depends_on('test');

        return;
    } # end ACTION_authortest()

    sub ACTION_distdir {
        my ($self) = @_;

        $self->depends_on('authortest');

        return $self->SUPER::ACTION_distdir();
    } # end ACTION_distdir

END_SUBCLASS

my $builder = $class->new(
    module_name         => 'MooseX::Accessors::ReadWritePrivate',
    dist_author         => 'Elliot Shank <perl@galumph.com>',
    license             => 'perl',
    dynamic_config      => 0,
    create_license      => 1,
    create_readme       => 0,
    create_packlist     => 1,
    sign                => 0,

    requires            => {
        'Moose'                     => 0.84,
        'Moose::Exporter'           => 0,
        'Moose::Role'               => 0,
        'Moose::Util::MetaRole'     => 0,
        'strict'                    => 0,
        'utf8'                      => 0,
        'version'                   => 0,
        'warnings'                  => 0,
    },

    build_requires      => {
        'Test::More'    => 0.72,
    },

    recommends          => {
    },

    meta_merge          => {
        configure_requires  => {
            'Module::Build' => 0.2808,
        },
        no_index    => {
            file        => [
                qw<
                >
            ],
            directory   => [
                qw<
                    xt
                >
            ],
        },
    },

    add_to_cleanup      => [
        qw<
            LICENSE
            MANIFEST
            MANIFEST.bak
            META.yml
            Makefile
            Makefile.old
            Makefile.PL
            pod2htm?.tmp
            pm_to_blib
            *.tmp
            *.bak
            */*.tmp
            */*.bak
            */*/*.tmp
            */*/*.bak
            */*/*/*.tmp
            */*/*/*.bak
            */*/*/*/*.tmp
            */*/*/*/*.bak
            MooseX-Accessors-ReadWritePrivate-*
        >
    ],
);


$builder->create_build_script();

# setup vim: set filetype=perl tabstop=4 softtabstop=4 expandtab :
# setup vim: set shiftwidth=4 shiftround textwidth=78 nowrap autoindent :
# setup vim: set foldmethod=indent foldlevel=0 :
